java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java

```diff
@@ -127,128 +132,131 @@ public void execute(EventMessage msg) {
+            });
     }
 
 
     /**
      * Performs minion registration.
-     *
      * @param minionId minion id
      * @param isSaltSSH true if a salt-ssh system is bootstrapped
      * @param actKeyOverride label of activation key to be applied to the system.
-     *                              If left empty, activation key from grains will be used.
+     *                       If left empty, activation key from grains will be used.
+     * @param machineId Machine Id of the minion
+     * @param saltbootInitrd saltboot_initrd, to be used for retail minions
      */
-    private void registerMinion(String minionId, boolean isSaltSSH,
-                                Optional<Long> saltSSHProxyId,
-                                Optional<String> actKeyOverride) {
-
-        // Match minions via their machine id
-        Opt.fold(SALT_SERVICE.getMachineId(minionId),
+    private void registerMinion(String minionId, boolean isSaltSSH, Optional<Long> saltSSHProxyId,
+                                Optional<String> actKeyOverride, String machineId, boolean saltbootInitrd) {
+        Optional<MinionServer> registeredMinionOpt = MinionServerFactory.findByMachineId(machineId);
+        Opt.consume(registeredMinionOpt,
             () -> {
-                LOG.info("Cannot find machine id for minion: " + minionId);
-                return false;
-            },
-            machineId -> {
-                boolean isReactivation = false;
-                ActivationKey ak = null;
-                ValueMap grains = new ValueMap(SALT_SERVICE.getGrains(minionId).orElseGet(HashMap::new));
-                Optional<String> managmentKeyLabel = getManagementKeyLabelFromGrains(grains);
-                if (managmentKeyLabel.isPresent()) {
-                    ak = ActivationKeyFactory.lookupByKey(managmentKeyLabel.get());
-                    if (ak != null && ak.getKickstartSession() != null) {
-                        ak.getKickstartSession().markComplete("Installation completed.");
-                    }
-                    if (ak == null) {
-                        LOG.info("Outdated Management Key defined for " + minionId + ": " + managmentKeyLabel.get());
-                    }
-                    else if (ak.getServer() == null) {
-                        LOG.error("Management Key is not a reactivation key: " + managmentKeyLabel.get());
-                    }
-                    else {
-                        isReactivation = true;
-                    }
-                }
+               if (!duplicateMinionNamePresent(minionId)) {
+                   ValueMap grains = new ValueMap(SALT_SERVICE.getGrains(minionId).orElseGet(HashMap::new));
+                   Optional<String> managmentKeyLabel = getManagementKeyLabelFromGrains(grains);
+                   Optional<User> creator = MinionPendingRegistrationService.getCreator(minionId);
+                   reactivateSystem(minionId, machineId, managmentKeyLabel);
```

Comment:
I have a hard time following if and why the logic has changed here.

Old code called `reactivateSystem` only if a reactivation key is used (a `managmentKeyLabel` is set in grains, it actually corresponds to an existing Activation Key, that Activation Key has an attached Server). It does not care whatever `duplicateMinionNamePresent` returns, in fact it does not call that method at all in case of a reactivation key (`isReactivation == true`).

New code seems to rule out that a reactivation may happen if a duplicate minion name is present - but that is a possibility.

Am I missing anything?

Reply:
Again correct. It should be better and correct now.